packages/lisk-transactions/src/base_transaction.ts

```diff
@@ -0,0 +1,263 @@
+		this.asset = rawTransaction.asset;
+		this.fee = new BigNum(rawTransaction.fee);
+		this.id = rawTransaction.id;
+		this.recipientId = rawTransaction.recipientId;
+		this.recipientPublicKey = rawTransaction.recipientPublicKey;
+		this.senderId = rawTransaction.senderId;
+		this.senderPublicKey = rawTransaction.senderPublicKey;
+		this.signature = rawTransaction.signature;
+		this.signatures = rawTransaction.signatures;
+		this.signSignature = rawTransaction.signSignature;
+		this.timestamp = rawTransaction.timestamp;
+		this.type = rawTransaction.type;
+	}
+
+	public abstract prepareTransaction(
+		passphrase: string,
+		secondPassphrase?: string,
+	): TransactionJSON;
+
+	public toJSON(): TransactionJSON {
+		const transaction = {
+			id: this.id,
+			amount: this.amount.toString(),
+			type: this.type,
+			timestamp: this.timestamp,
+			senderPublicKey: this.senderPublicKey,
+			senderId: this.senderId,
+			recipientId: this.recipientId,
+			recipientPublicKey: this.recipientPublicKey,
+			fee: this.fee.toString(),
+			signatures: this.signatures,
+			asset: this.asset,
+		};
+
+		if (!(typeof this.signature === 'string' && this.signature.length > 0)) {
+			return transaction;
+		}
+
+		const singleSignedTransaction = {
+			...transaction,
+			signature: this.signature,
+		};
+
+		if (
+			!(typeof this.signSignature === 'string' && this.signSignature.length > 0)
+		) {
+			return singleSignedTransaction;
+		}
+
+		const signedTransaction = {
+			...singleSignedTransaction,
+			signSignature: this.signSignature,
+		};
+
+		return signedTransaction;
+	}
+
+	public getBytes(): Buffer {
+		const { signature, signSignature } = this.toJSON();
+		const transactionType = Buffer.alloc(BYTESIZES.TYPE, this.type);
+		const transactionTimestamp = Buffer.alloc(BYTESIZES.TIMESTAMP);
+		transactionTimestamp.writeIntLE(this.timestamp, 0, BYTESIZES.TIMESTAMP);
+
+		const transactionSenderPublicKey = cryptography.hexToBuffer(
+			this.senderPublicKey,
+		);
+
+		const transactionRecipientID = this.recipientId
+			? cryptography.bigNumberToBuffer(
+					this.recipientId.slice(0, -1),
+					BYTESIZES.RECIPIENT_ID,
+			  )
+			: Buffer.alloc(BYTESIZES.RECIPIENT_ID);
+
+		const amountBigNum = new BigNum(this.amount);
+
+		const transactionAmount = amountBigNum.toBuffer({
+			endian: 'little',
+			size: BYTESIZES.AMOUNT,
+		});
+
+		const transactionSignature = signature
+			? cryptography.hexToBuffer(signature)
+			: Buffer.alloc(0);
+
+		const transactionSecondSignature = signSignature
+			? cryptography.hexToBuffer(signSignature)
+			: Buffer.alloc(0);
+
+		return Buffer.concat([
+			transactionType,
+			transactionTimestamp,
+			transactionSenderPublicKey,
+			transactionRecipientID,
+			transactionAmount,
+			transactionSignature,
+			transactionSecondSignature,
+		]);
+	}
+
+	public abstract containsUniqueData(): boolean;
+
+	public validate(): ValidateReturn {
+		const transaction = this.toJSON();
+
+		if (!TRANSACTION_TYPES.includes(transaction.type)) {
```

Comment:
if we validate type by known ID, I think it will limit us not being able to add custom transaction.
Maybe type check should be done in extended class?